src/lib.rs

```diff
@@ -947,115 +950,129 @@ impl Executor {
-        }
-
-        let (builders, other) = match processes {
-            0 => timely::CommunicationConfig::Process(threads),
-            _ => {
-                let addresses = addresses.unwrap_or_else(|| {
-                    let mut local_addrs = Vec::new();
-                    for index in 0..processes {
-                        local_addrs.push(format!("localhost:{}", 2101 + index));
+        py.allow_threads(move || {
+            let (builders, other) = match processes {
+                0 => timely::CommunicationConfig::Process(threads),
+                _ => {
+                    let addresses = addresses.unwrap_or_else(|| {
+                        let mut local_addrs = Vec::new();
+                        for index in 0..processes {
+                            local_addrs.push(format!("localhost:{}", 2101 + index));
+                        }
+                        local_addrs
+                    });
+                    timely::CommunicationConfig::Cluster {
+                        threads,
+                        process,
+                        addresses,
+                        report: false,
+                        log_fn: Box::new(|_| None),
                     }
-                    local_addrs
-                });
-                timely::CommunicationConfig::Cluster {
-                    threads,
-                    process,
-                    addresses,
-                    report: false,
-                    log_fn: Box::new(|_| None),
                 }
             }
-        }
-        .try_build()
-        .unwrap();
-
-        let interrupt_flag_for_worker_threads = interrupt_flag.clone();
-        let guards = timely::execute::execute_from(
-            builders,
-            other,
-            WorkerConfig::default(),
-            move |worker| {
-                let mut all_pumps = Vec::new();
-                let mut all_probes = Vec::new();
-
-                Python::with_gil(|py| {
-                    // These borrows should be safe because nobody else
-                    // will have a mut ref. Although they might have
-                    // shared refs.
-                    let self_ = self_.as_ref(py).borrow();
-                    for dataflow in &self_.dataflows {
-                        let dataflow = dataflow.as_ref(py).borrow();
-                        let (pump, probe) = build_dataflow(worker, py, &dataflow);
-                        all_pumps.extend(pump);
-                        all_probes.push(probe);
+            .try_build()
+            .map_err(PyRuntimeError::new_err)?;
+
+            let should_shutdown = Arc::new(AtomicBool::new(false));
+            let should_shutdown_w = should_shutdown.clone();
+            let should_shutdown_p = should_shutdown.clone();
+            // We can drop these if we want to use nightly
+            // Thread::is_running
+            // https://github.com/rust-lang/rust/issues/90470
```

Comment:
When you were deciding on this approach over the `is_running`, what were the trade-offs?

Reply:
We can't use `is_running` without flipping our whole Rust tool chain over to using nightly builds. I'm concerned about the stability guarantees of that. So this wasn't really an option, but I wanted to note it in case it does become stable eventually.